### PR TITLE
Video: implement output resampling (upscaling/downscaling) methods

### PR DESCRIPTION
**Pull request by Filoppi**
Follow up to:
https://github.com/dolphin-emu/dolphin/pull/11850

This PR implements custom upscaling/downscaling techniques in our default post process pass.
By default no custom resampling is done, Dolphin just falls back on the GPU hardware (bi)linear sampling (which has no effect unless the emulation output (XFB) resolution doesn't match the window resolution).
If the user selects any resampling method other than default, then the default post process pass will be executed, and it will do some custom resampling (before any other user selected post process), with gamma correction (colors are converted to linear space before resampling).

I have implemented a multitude of algorithms, all of them have been rigorously tested and checked against each other and against many other code bases (e.g. Cemu, Yuzu, DXVK, github random samples, ...).
There's no "best" resampling method so each user can select what they prefer depending on the case.
image

Given that resampling is optional and that Dolphin doesn't have major performance issues, the shaders might not be as optimized as they could, but for now I'd say there isn't much to worry about.
Everybody is welcome to add new resampling methods, and the Catmull-Rom is basically a template for any other bicubic algorithms.

Upscaling (from 1x):
Bilinear
Bicubic
Hermite
Catmull-Rom
NearestNeighbor
SharpBilinear

Downscaling (from 8x):
Bilinear
Bicubic
Hermite
Catmull-Rom
NearestNeighbor
SharpBilinear

New Box Resampling (for downscaling) by @Sam-Belliveau.
Box
Catmull-Rom
Box might look a bit less "sharp", but it's extremely termporal stable if the source resolution is high enough.

More samples:
https://github.com/dolphin-emu/dolphin/pull/12018

### CODE REVIEW

**Comment by AdmiralCurtiss** [comment]
I can't say much about the shader code, but this is a really good improvement to image quality when downsampling, it's always been pretty terrible before. This actually makes resolutions above the screen size worth using.

**Review by iwubcode — APPROVED** [review]

**Review by Sam-Belliveau — APPROVED** [review]

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp [hunk 1]
```diff
@@ -453,6 +489,18 @@ void EnhancementsWidget::AddDescriptions()
       "scaling filter selected by the game.<br><br>Any option except 'Default' will alter the look "
       "of the game's textures and might cause issues in a small number of "
       "games.<br><br><dolphin_emphasis>If unsure, select 'Default'.</dolphin_emphasis>");
+  static const char TR_OUTPUT_RESAMPLING_DESCRIPTION[] = QT_TR_NOOP(
+      "Affects how the game output image is upscaled or downscaled to the window resolution.<br>"
+      "\"Default\" will rely on the GPU internal bilinear sampler which isn't gamma corrected."
+      "<br>\"Bilinear\" (gamma corrected) is a good compromise between quality and performance."
+      "<br>\"Bicubic\" is smoother than \"Bilinear\"."
+      "<br>\"Hermite\" might offer the best quality when upscaling,"
+      " at a slightly bigger perform cost.<br>\"Catmull-Rom\" is best for downscaling."
```
Comment (MayImilae) [inline Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:498]: `"<br>\"Hermite\" might offer the best quality when upscaling,"
      " at a slightly bigger perform cost.<br>\"Catmull-Rom\" is best for downscaling."`

Seems you meant to put Catmull-Rom on a new line?

Also, I'd suggest "may" rather than "might" here. Though honestly in general I'd prefer to avoid indefinites in the language of these tooltips but I can't actually test it right now so, whatever I guess.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp [hunk 2]
```diff
@@ -453,6 +489,18 @@ void EnhancementsWidget::AddDescriptions()
...
+      "<br>\"Nearest Neighbor\" doesn't do any resampling, select if you like a pixelated look."
```
Comment (MayImilae) [inline Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:499]: So um, why is nearest neighbor even here? Like, I know it's cheap and easy, but it has a lot of downsides and artifacts.... and we have sharp bilinear. We should be pushing anyone who likes the pixelated look toward sharp bilinear. I don't think we should even be giving Nearest Neighbor as an option to select in this menu.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp [hunk 3]
```diff
@@ -453,6 +489,18 @@ void EnhancementsWidget::AddDescriptions()
...
+      "<br>\"Sharp Bilinear\" works best with 2D games at low resolutions, use if you like a sharp"
```
Comment (MayImilae) [inline Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:500]: "sharp look" is a really bad way to put this. This is the option users should take if they want a pixelated look, not nearest neighbor.

And since it's all about pixel clarity, it provides basically zero benefits at higher resolutions. Saying it is a "sharp look" will make users think it's a sharpening filter! This needs to be changed.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp [hunk 4]
```diff
@@ -453,6 +489,18 @@ void EnhancementsWidget::AddDescriptions()
       "scaling filter selected by the game.<br><br>Any option except 'Default' will alter the look "
       "of the game's textures and might cause issues in a small number of "
       "games.<br><br><dolphin_emphasis>If unsure, select 'Default'.</dolphin_emphasis>");
+  static const char TR_OUTPUT_RESAMPLING_DESCRIPTION[] = QT_TR_NOOP(
+      "Affects how the game output image is upscaled or downscaled to the window resolution.<br>"
+      "\"Default\" will rely on the GPU internal bilinear sampler which isn't gamma corrected."
+      "<br>\"Bilinear\" (gamma corrected) is a good compromise between quality and performance."
+      "<br>\"Bicubic\" is smoother than \"Bilinear\"."
```
Comment (MayImilae) [inline Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:496]: "smoother" doesn't really mean anything here. In practice, bicubic is _slightly_ sharper than bilinear.

Honestly I'm not sure why you even are offering bicubic when it's not a substantial improvement over bilinear - it seems like an easy cut to simplify this drop down.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp [hunk 5]
```diff
@@ -453,6 +489,18 @@ void EnhancementsWidget::AddDescriptions()
...
+      "<br>\"Hermite\" might offer the best quality when upscaling,"
+      " at a slightly bigger perform cost.<br>\"Catmull-Rom\" is best for downscaling."
+      "<br>\"Nearest Neighbor\" doesn't do any resampling, select if you like a pixelated look."
+      "<br>\"Sharp Bilinear\" works best with 2D games at low resolutions, use if you like a sharp"
+      " look."
+      "<br>\"Box Resampling\" is most expensive but also most accurate downscaling method."
```
Comment (MayImilae) [inline Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:502]: A comparison to SSAA may be useful here. After all, when we're at the stage of being fine with expensive downsampling, SSAA comes into the picture, and will be the primary competitor to this Box Resampling option.

**Review by MayImilae — CHANGES_REQUESTED** [review]
So I can't actually test this right now, on account of my computer being in pieces atm. But I can just review the UX!

Also it feels really weird to have gamma correction in the dropdown for scaling methods but, if that's how the hardware works then I guess that's that. If you could give me a quick explanation on that I'd appreciate it.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 1 above)
Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:498]: Sam has already changed all these tooltips for a follow up PR:
https://github.com/dolphin-emu/dolphin/pull/12018

I purposely used indefinites because it's subjective and it depends on a case by case basis. I didn't want people to say: "oh, it says it looks better thus I will use this."
My intention was for them to evaluate each themselves.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 2 above)
Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:499]: I don't see any reason to not expose it. We can put it last in the order?

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 3 above)
Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:500]: Please write down a full description I could use for it, I didn't investigate much the use cases for the sharp bilinear filtering.
Whatever you come up with, I will use, I don't particularly care about it.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 2 above)
Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:499]: Since you are adding something new, the opposite approach should be used, imo. The addition needs to be justified. In this case, why would we expose a setting we don't want users to use? That we know is worse than another option we have? That sounds like a perfect INI only option to me!

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 3 above)
Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:500]: Um, I'll do something off the top of my head here. References are https://github.com/dolphin-emu/dolphin/pull/9860 and https://dolphin-emu.org/blog/2023/05/21/dolphin-progress-report-february-march-april-2023/#new-post-processing-shaders

> Sharp Bilinear offers upscaling with sharp rendered pixels while still maintaining consistent pixel sizes. Use with native internal resolution for a pixelated look.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 1 above)
Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:498]: Fair enough. Well, I'd still like for you to move from might to may though.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 2 above)
Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:499]: @AdmiralCurtiss do you agree that not having Nearest Neighbor is good and that Sharp Bilinear is better for all cases? If so, I'm on with not exposing it.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 5 above)
Comment (Sam-Belliveau) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:502]: The main difference between box/area resampling and SSAA is that SSAA is not a window resampler. The best example is, say i have a really weird TV, it's 1366x768, and I want the best pixel quality, i can render it at 8k, and Box Resampling will fit the pixels optimally into that weird 1366x768 screen so it looks smooth. This is a weird example and I'm not sure how to fit it into the description.

Comment (v1993) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:502]: Practically speaking (and weird resolutions discussed above aside), SSAA at the same multiplier as `(internal resolution / display resolution)^2` (e.g. 1080p 4x SSAA vs 4k no AA) will produce superior results because of more complex (compared to axis-aligned grid that high internal resolution gives) sampling pattern typically used by the driver. However, some games suffer from artifacts when SSAA is used, leaving higher internal resolutions as the only viable options.

The bottom line is: high-quality resampling + high internal resolution gives lower image quality at resource consumption very close to comparable SSAA mode, but higher compatibility with games.

Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:502]: Resolution scaling is also beneficial when your display resolution or window doesn't match the rendering resolution, which is basically all the times.

Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: @v1993 what games have issues with SSAA?

Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: I messed around with SSAA a bit. It's my understanding that it works so that the game output resolution stays the same, but internally each pixel was an average of x (e.g. 8) samples instead of 1.

This is great when playing at a low internal resolution (e.g. wii/gc native) in a small or big game window,
but it doesn't actually help at all when playing at a high internal resolution in a window of a <= resolution, because the downscaling was handled poorly, thus all the extra detail would end up lost.
Additionally both MSAA and SSAA cause artifacts in multiple games (e.g. Super Mario Galaxy, Super Mario Sunshine).

Basically SSAA is a "resolution quality scaler" that applies within the rendering phase, while my resampling just deals with the game->window resolution conversion. They aren't really competitors, and 8x SSAA does not look better than 8x internal resolution + area/box downsampling.
As shown in the examples here, SSAA presumably loses a lot of potentially additional detail due it being stored in a lower resolution image.

Native:
Dolphin_6NrCkMR9Zq

Native + 8x SMAA:
Dolphin_F9b2JZ8iNs

8x Native res + Area Resampling:
Dolphin_x2xVePrD8p

**Comment by Filoppi** [comment]
I've merged SamB changes into the branch and solved all the remaining issues mentioned above.
I've unexposed nearest neighbor resampling from users and added a comparison with SSMAA in the tooltip.

New tooltip:
Dolphin_g25nI1o9lO

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 5 above)
Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: > I messed around with SSAA a bit.

So for best results, we've always told users to configure the Internal Resolution to match their screen resolution (or at least as close as possible), then apply SSAA. Is that what you did when you were experimenting with it?

Comment (v1993) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: You have an major error in your test: 8x SSAA gives you four samples per pixel, 8x IR gives you 8*8=64 samples per pixel - sample count increases exponentially as resolution grows. To approximate 8x SSAA with IRs you should use `sqrt(8)=2.82...≅3`x IR.

The idea behind comparing box resampling with SSAA is that both effectively result in multiple samples of underlying primitives, providing "true" anti-aliasing. Downsampling has the benefit of higher compatibility (as you've noted for two Mario games, see https://wiki.dolphin-emu.org/index.php?title=Category:Anti-Aliasing_(Config_Required) for some more examples) and while SSAA takes advantage of non-aligned-grid sampling pattern, which allows to eliminate aliasing at much lower performance cost.

Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: @MayImilae that's a good suggestion to give to users. I play at 1440p and could barely notice any added quality from SSAA (it actually caused glitches on the pipes in the screenshot), and given I can play at 60fps with a 8x resolution mutliplier + area resampling, I'd personally see no use in SSAA, but I guess it depends on the user and the hardware.

Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: @v1993 good point, I didn't think about the different ways the multipliers are applied.
I supposed when comparing x8 SSAA and ~x3 IR, the result is pretty similar. There's not much need in defining which one would generally look best in that case as they still have different purposes.

Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: 3x IR still looks a lot better than 8x SSAA
Dolphin_p6C5Ftgc6I
Dolphin_nz27R3q0zm
Dolphin_sD6j6VcTuy

File: Data/Sys/Shaders/default_pre_post_process.glsl [hunk 6]
```diff
@@ -188,12 +188,14 @@ float4 AreaSampling(float3 uvw, float gamma)
...
 	float2 source_size = GetResolution();
+	float2 inverted_source_size = GetInvResolution();
 	float2 inverted_target_size = GetInvWindowResolution();
 
 	// Determine the range of the source image that the target pixel will cover.
 	// We shift by one output pixel because that's a prerequisite of the algorithm.
 	float2 range = source_size * inverted_target_size;
-	float2 beg = (uvw.xy - inverted_target_size) * source_size;
+	// NOTE: these offsets make the output shift by ~half a pixel when upscaling
+	float2 beg = (uvw.xy - (inverted_target_size * 0.5) - (inverted_source_size * 0.5)) * source_size;
```
Comment (Sam-Belliveau) [inline Data/Sys/Shaders/default_pre_post_process.glsl:198]: `float2 beg = (uvw.xy - (inverted_target_size * 0.5)) * source_size - float2(0.5, 0.5);`

might make more sense here.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 5 above)
Comment (v1993) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: Now *that* is impressive. While box resampled picture has a bit more aliasing, many would consider it much more desirable for significantly cleaner look (and you can pump that into some post-processor doing SMAA to smooth out remaining jaggies). Have you tried this with more games, especially those not known for issues with MSAA/SSAA? I'd expect higher IRs to be superior for games utilizing render-to-texture effects, although I'm not sure by how much.

Also, may I ask what card/drivers are those? Exact SSAA/MSAA implementation varies per vendor and possibly even somewhat per graphics API.

Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: DX12 RTX 3090

**Comment by Sam-Belliveau** [comment]
I dont know how much of an improvement the current tooltip is over how it was before?

Before:
image

After:
image

We might wanna think about how to make it clearer?

**Review by Sam-Belliveau — APPROVED** [review]
Looked over the GLSL code and the tool tip.

LGTM

**Comment by AdmiralCurtiss** [comment]
Looks like I introduced a conflict here when I switched the tooltip for the Color Correction to our fancy ones. Can you rebase?

**Comment by Filoppi** [comment]
@AdmiralCurtiss done.

**Comment by Frieslover69** [comment]
> I've unexposed nearest neighbor resampling from users and added a comparison with SSMAA in the tooltip.

Hi @Filoppi, first of all thank you for adding this awesome feature! Being able to select the output filtering modes is really something every emulator should have imho.
Speaking of basic filtering modes, is there a reason why you decided to remove the nearest neighbor resampling from the dropdown? I know some could argue that sharp bilinear or area sampling would make it obsolete, but some users (such as myself 😄) might want to enjoy the games on nearest neighbor (a.k.a. no filtering). The reason for this is that area sampling still introduces some "ghosting" (for a lack of better words) to the pixels of the image. Whereas nearest neighbor filtering truly gives you the pure unadulterated expierience of absolute sharp edges in all cases. Having the option to basically selecty no filtering at all would make sense for such a dropdown imho.

So my suggestion would be to maybe add the nearest neighbor filtering back to the menu? All the pixel purists out there would greatly appreaciate it, I'm sure! 🙏

Best regards 😊

**Comment by mbc07** [comment]
@Frieslover69 there's an explanation in [this blog post](https://dolphin-emu.org/blog/2023/05/21/dolphin-progress-report-february-march-april-2023/#new-post-processing-shaders) regarding the downsides of a "pure" Nearest Neighbor filter in older consoles like the GC/Wii and that's why we're not providing it as an option...

**Comment by Frieslover69** [comment]
@mbc07 Thanks for the reply! Yeah, I'm aware of the downsides of nearest neighbor. But some things might be considered before ultimately abandoning that feature:
1. The flickering gets less noticable at larger upscales. At some point it becomes even unnoticable! A bump to let's say 150% of the original resolution leads to noticable flickering. That is because the thickness of the columns (as it is called in the blog) alternates between 1 or 2 pixels. One column is twice as thick as the other! But if you bump the resolution to 250% using NN, then the thickness of the column alternates between 2 or 3 pixels. So the "bad" columns are only a third off the good column size. In that case, the flickering is less noticable. Especially on 1440p monitors you'd usually use the NN to scale the original image to even more than 250%, resulting in even less flickering.
So TL;DR: flickering is no big deal on higher upscales (i.e. bumping from original GC res to 1440p).
2. Other filters have their downsides too... Some of them might give you an undesirable soft look, but they've been left in nonetheless.
3. The simple option to have NN wouldn't hurt. If a user is unhappy with the look, they can just choose to select another filter from the list. 😄

So... do you think NN has a chance of getting accepted if someone submits a pull-request? The code for the filter is still in the default_pre_post_process.glsl file, so adding it back would only need a relatively small change. And the first post even says everyone is free to add their own filters. 😊

**Comment by Filoppi** [comment]
@Frieslover69 leaving nearest  neighbor has been discussed at length and we decided against (I'm not as against it as others). We are aware there are some positives of keeping it, but at the same time, it's also extremely easy for users to re-enable it by just commenting out in `default_pre_post_process.glsl` (the code is still there, just disabled).
The main drawback of nearest neighbor is that GC and Wii games resolution is almost always scaled from its native aspect ratio to ~4:3 or ~16:9, thus nearest neighbor doesn't make much sense given pixels weren't squared. I suggest using sharp bilinear to achieve the best results.

These kind of discussions are probably best had on discord, join the devs channel there if you want to have further discussion, or just make a PR to add back nearest neighbor.
